### PR TITLE
[skia] Adding support for other graphics APIs

### PR DESCRIPTION
**Pull request by dave-juicelabs**
Updating skia to add support for other APIs other than OpenGL.

- #### What does your PR fix?  

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  Unchanged from previous. CI baseline is updated.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by JonLiu1993** [comment]
Please run command "./vcpkg x-add-version skia --overwrite-version" then commit the changes again

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Please run command "./vcpkg x-add-version skia --overwrite-version" then commit the changes again

File: ports/skia/CONTROL [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Source: skia
-Version: 2020-05-18
+Version: 2021-08-30
 Port-Version: 4
```
Comment (JonLiu1993) [inline ports/skia/CONTROL:3]:
```suggestion
Port-Version: 5
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Please run command "./vcpkg x-add-version skia --overwrite-version" then commit the changes again

File: versions/s-/skia.json [hunk 2]
```diff
@@ -1,7 +1,17 @@
...
-      "git-tree": "eea166718304da2660d795937fe2528dbd2909cc",
+      "git-tree": "0eeadf5aad450fdf2b4f63a199cc9c9221127830",
+      "version-string": "2021-08-30",
+      "port-version": 5
+    },
+    {
+      "git-tree": "bf350ede3bd4cfe0d7de80ab835348414f7313f5",
+      "version-string": "2021-08-30",
+      "port-version": 4
+    },
```
Comment (JonLiu1993) [inline versions/s-/skia.json:12]:
```suggestion
```

File: versions/s-/skia.json [hunk 3]
```diff
@@ -1,7 +1,17 @@
...
+    {
+      "git-tree": "1d80dcb05c6608eaf63cfb07380f82c602e2cb56",
```
Comment (JonLiu1993) [inline versions/s-/skia.json:14]:
```suggestion
      "git-tree": "eea166718304da2660d795937fe2528dbd2909cc",
```

File: versions/s-/skia.json (hunk 2 above)
Comment (JonLiu1993) [inline versions/s-/skia.json:12]:
```suggestion

```

File: ports/skia/vcpkg.json [hunk 4]
```diff
@@ -1,7 +1,6 @@
 {
   "name": "skia",
-  "version-string": "2020-05-18",
-  "port-version": 5,
+  "version-string": "2021-08-06",
```
Comment (JonLiu1993) [inline ports/skia/vcpkg.json:3]:
```suggestion
  "version-date": "2021-08-06",
```

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Could you please take a look:
````
[799/799] link skia.dll skia.dll.lib skia.dll.pdb
FAILED: skia.dll skia.dll.lib skia.dll.pdb 
"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC/Tools/MSVC/14.29.30133/bin/HostX64/x64/link.exe" /nologo /IMPLIB:./skia.dll.lib /DLL /OUT:./skia.dll /PDB:./skia.dll.pdb @./skia.dll.rsp
   Creating library ./skia.dll.lib and object ./skia.dll.exp
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_blob_create referenced in function "class std::unique_ptr<struct hb_blob_t,struct SkFunctionWrapper<void __cdecl(struct hb_blob_t *),&void __cdecl hb_blob_destroy(struct hb_blob_t *)> > __cdecl to_blob(class sk_sp<class SkData>)" (?to_blob@@YA?AV?$unique_ptr@Uhb_blob_t@@U?$SkFunctionWrapper@$$A6AXPEAUhb_blob_t@@@Z$1?hb_blob_destroy@@YAX0@Z@@@std@@V?$sk_sp@VSkData@@@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_blob_destroy referenced in function "public: __cdecl <lambda_1b826c87cf9bf7a1c0fcf379fd41c482>::operator()(void const *,void *)const " (??R<lambda_1b826c87cf9bf7a1c0fcf379fd41c482>@@QEBA@PEBXPEAX@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_blob_get_data referenced in function "class sk_sp<class SkData> __cdecl to_data(class std::unique_ptr<struct hb_blob_t,struct SkFunctionWrapper<void __cdecl(struct hb_blob_t *),&void __cdecl hb_blob_destroy(struct hb_blob_t *)> >)" (?to_data@@YA?AV?$sk_sp@VSkData@@@@V?$unique_ptr@Uhb_blob_t@@U?$SkFunctionWrapper@$$A6AXPEAUhb_blob_t@@@Z$1?hb_blob_destroy@@YAX0@Z@@@std@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_set_add referenced in function "public: __cdecl <lambda_7aa85c9c4411a3e1c4254e48c2302f10>::operator()(unsigned int)const " (??R<lambda_7aa85c9c4411a3e1c4254e48c2302f10>@@QEBA@I@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_face_create referenced in function "class sk_sp<class SkData> __cdecl subset_harfbuzz(class sk_sp<class SkData>,class SkPDFGlyphUse const &,int)" (?subset_harfbuzz@@YA?AV?$sk_sp@VSkData@@@@V1@AEBVSkPDFGlyphUse@@H@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_face_destroy referenced in function "public: void __cdecl SkFunctionWrapper<void __cdecl(struct hb_face_t *),&void __cdecl hb_face_destroy(struct hb_face_t *)>::operator()<struct hb_face_t * &>(struct hb_face_t * &)const " (??$?RAEAPEAUhb_face_t@@@?$SkFunctionWrapper@$$A6AXPEAUhb_face_t@@@Z$1?hb_face_destroy@@YAX0@Z@@QEBAXAEAPEAUhb_face_t@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_face_reference_blob referenced in function "class sk_sp<class SkData> __cdecl subset_harfbuzz(class sk_sp<class SkData>,class SkPDFGlyphUse const &,int)" (?subset_harfbuzz@@YA?AV?$sk_sp@VSkData@@@@V1@AEBVSkPDFGlyphUse@@H@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset_input_create_or_fail referenced in function "class sk_sp<class SkData> __cdecl subset_harfbuzz(class sk_sp<class SkData>,class SkPDFGlyphUse const &,int)" (?subset_harfbuzz@@YA?AV?$sk_sp@VSkData@@@@V1@AEBVSkPDFGlyphUse@@H@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset_input_destroy referenced in function "public: void __cdecl SkFunctionWrapper<void __cdecl(struct hb_subset_input_t *),&void __cdecl hb_subset_input_destroy(struct hb_subset_input_t *)>::operator()<struct hb_subset_input_t * &>(struct hb_subset_input_t * &)const " (??$?RAEAPEAUhb_subset_input_t@@@?$SkFunctionWrapper@$$A6AXPEAUhb_subset_input_t@@@Z$1?hb_subset_input_destroy@@YAX0@Z@@QEBAXAEAPEAUhb_subset_input_t@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset_input_glyph_set referenced in function "class sk_sp<class SkData> __cdecl subset_harfbuzz(class sk_sp<class SkData>,class SkPDFGlyphUse const &,int)" (?subset_harfbuzz@@YA?AV?$sk_sp@VSkData@@@@V1@AEBVSkPDFGlyphUse@@H@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset_input_set_drop_hints referenced in function "public: static class std::unique_ptr<struct hb_face_t,struct SkFunctionWrapper<void __cdecl(struct hb_face_t *),&void __cdecl hb_face_destroy(struct hb_face_t *)> > __cdecl SkPDFHarfBuzzSubset<struct hb_subset_input_t *,void>::Make(struct hb_subset_input_t *,struct hb_face_t *)" (?Make@?$SkPDFHarfBuzzSubset@PEAUhb_subset_input_t@@X@@SA?AV?$unique_ptr@Uhb_face_t@@U?$SkFunctionWrapper@$$A6AXPEAUhb_face_t@@@Z$1?hb_face_destroy@@YAX0@Z@@@std@@PEAUhb_subset_input_t@@PEAUhb_face_t@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset_input_set_retain_gids referenced in function "public: static class std::unique_ptr<struct hb_face_t,struct SkFunctionWrapper<void __cdecl(struct hb_face_t *),&void __cdecl hb_face_destroy(struct hb_face_t *)> > __cdecl SkPDFHarfBuzzSubset<struct hb_subset_input_t *,void>::Make(struct hb_subset_input_t *,struct hb_face_t *)" (?Make@?$SkPDFHarfBuzzSubset@PEAUhb_subset_input_t@@X@@SA?AV?$unique_ptr@Uhb_face_t@@U?$SkFunctionWrapper@$$A6AXPEAUhb_face_t@@@Z$1?hb_face_destroy@@YAX0@Z@@@std@@PEAUhb_subset_input_t@@PEAUhb_face_t@@@Z)
pdf.SkPDFSubsetFont.obj : error LNK2019: unresolved external symbol hb_subset referenced in function "public: static class std::unique_ptr<struct hb_face_t,struct SkFunctionWrapper<void __cdecl(struct hb_face_t *),&void __cdecl hb_face_destroy(struct hb_face_t *)> > __cdecl SkPDFHarfBuzzSubset<struct hb_subset_input_t *,void>::Make(struct hb_subset_input_t *,struct hb_face_t *)" (?Make@?$SkPDFHarfBuzzSubset@PEAUhb_subset_input_t@@X@@SA?AV?$unique_ptr@Uhb_face_t@@U?$SkFunctionWrapper@$$A6AXPEAUhb_face_t@@@Z$1?hb_face_destroy@@YAX0@Z@@@std@@PEAUhb_subset_input_t@@PEAUhb_face_t@@@Z)
.\skia.dll : fatal error LNK1120: 13 unresolved externals
ninja: build stopped: subcommand failed.
````

**Comment by JonLiu1993** [comment]
@damcclos ,It seems that the CI error has not changed from last time，after finish your suggesstion changes,please run command "./vcpkg x-add-version skia --overwrite-version" then commit the changes again

**Comment by JonLiu1993** [comment]
@dave-juicelabs ，Is this PR being worked on?

**Comment by dave-juicelabs** [comment]
@JonLiu1993 Yes, I will be updating this PR.

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,I tested feature skia[*]:x64-windows and skia[*]:x64-windows-static locally but failed :
````
ERROR at //BUILD.gn:17:3: Unable to load "F:/test-pr/skia/vcpkg/buildtrees/skia/src/fa8ee834ed-4d075674a2.clean/third_party/externals/dawn/scripts/dawn_features.gni".
  import("third_party/externals/dawn/scripts/dawn_features.gni")
  ^------------------------------------------------------------
````

**Comment by dave-juicelabs** [comment]
Tested all the different features with https://github.com/dave-juicelabs/skia-vcpkg-integration.

**Comment by JackBoosY** [comment]
Please get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=62163&view=artifacts&pathAsName=false&type=publishedArtifacts).

**Comment by dave-juicelabs** [comment]
Separated gn update from this PR, #21114 . Failures will be fixed there first before this should be merged.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Could you please take a look:
````
[20/881] compile ../src/fa8ee834ed-2b62b425cf.clean/src/ports/SkFontMgr_fontconfig.cpp
FAILED: obj/src/ports/fontmgr_fontconfig.SkFontMgr_fontconfig.o 
c++ -MD -MF obj/src/ports/fontmgr_fontconfig.SkFontMgr_fontconfig.o.d -DSK_R32_SHIFT=16 -DSK_ENABLE_SKSL -DSK_GAMMA_APPLY_TO_A8 -DSK_ALLOW_STATIC_GLOBAL_INITIALIZERS=1 -DGR_TEST_UTILS=1 -DSKIA_IMPLEMENTATION=1 -I../src/fa8ee834ed-2b62b425cf.clean -Wno-attributes -fstrict-aliasing -fPIC -fvisibility=hidden -g -Wall -Wextra -Winit-self -Wpointer-arith -Wsign-compare -Wvla -Wno-deprecated-declarations -Wno-maybe-uninitialized -Wno-psabi -Wno-unused-parameter -std=c++17 -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -Wnon-virtual-dtor -Wno-noexcept-type -Wno-redundant-move -c ../src/fa8ee834ed-2b62b425cf.clean/src/ports/SkFontMgr_fontconfig.cpp -o obj/src/ports/fontmgr_fontconfig.SkFontMgr_fontconfig.o
../src/fa8ee834ed-2b62b425cf.clean/src/ports/SkFontMgr_fontconfig.cpp:27:10: fatal error: fontconfig/fontconfig.h: No such file or directory
   27 | #include <fontconfig/fontconfig.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
````

**Comment by dave-juicelabs** [comment]
Builds on #21114, which should be merged first.

**Comment by dave-juicelabs** [comment]
Now depends on #21498

**Comment by JackBoosY** [comment]
Please merge to master first.

File: ports/skia/portfile.cmake [hunk 5]
```diff
@@ -1,12 +1,13 @@
 vcpkg_fail_port_install(
     ON_ARCH "x86" "arm" "arm64"
-    ON_TARGET "UWP" "LINUX" "ANDROID" "FREEBSD")
+    ON_TARGET "UWP" "ANDROID" "FREEBSD")
```
Comment (JackBoosY) [inline ports/skia/portfile.cmake:3]: We don't need `vcpkg_fail_port_install` now, the keyword `supports` in vcpkg.json do the same job.
Please remove this.

File: ports/skia/vcpkg.json [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
-  "port-version": 6,
+  "version-date": "2020-11-30",
   "description": [
     "Skia is an open source 2D graphics library which provides common APIs that work across a variety of hardware and software platforms.",
     "It serves as the graphics engine for Google Chrome and Chrome OS, Android, Mozilla Firefox and Firefox OS, and many other products.",
     "Skia is sponsored and managed by Google, but is available for use by anyone under the BSD Free Software License. While engineering of the core components is done by the Skia development team, we consider contributions from any source."
   ],
   "homepage": "https://skia.org",
-  "supports": "x64 & (osx | windows)",
+  "supports": "x64",
```
Comment (JackBoosY) [inline ports/skia/vcpkg.json:10]:
```suggestion
  "supports": "x64 & !(uwp | android)",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JonLiu1993 — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
Looks okay to me.

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Thanks for your hard working, I tested all features locally but failed:

F:\test-pr\skia\vcpkg\buildtrees\skia\build-x64-windows-dbg-out.log:
````
ninja: Entering directory `F:/test-pr/skia/vcpkg/buildtrees/skia/x64-windows-dbg'

````
F:\test-pr\skia\vcpkg\buildtrees\skia\build-x64-windows-dbg-err.log:
````
ninja: error: obj/third_party/spirv-cross/spirv_cross.ninja:11: multiple rules generate obj/third_party/externals/spirv-cross/spirv_cross.spirv_cfg.obj [-w dupbuild=err]

````
Please use vcpkg spirv-cross port

**Comment by dave-juicelabs** [comment]
@JonLiu1993 What command line did you use to generate that error?

**Comment by JonLiu1993** [comment]
> @JonLiu1993 What command line did you use to generate that error?

./vcpkg install skia[*]:x64-windows

**Comment by JonLiu1993** [comment]
@dave-juicelabs, I tested all features again on different machines and failed because of the same error.

**Comment by dave-juicelabs** [comment]
I've been caught up in other work but I have been able to reproduce this error. I will be getting back to this shortly.

**Comment by JonLiu1993** [comment]
@dave-juicelabs , any updates?

**Comment by dave-juicelabs** [comment]
Not sure what the failure is or how to debug it at the moment.

**Comment by Thomas1664** [comment]
> Not sure what the failure is or how to debug it at the moment.

Please make sure you're up to date with master.

**Comment by JackBoosY** [comment]
The regression is a python3 issue which is not related to this PR.

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/23525.

**Comment by JackBoosY** [comment]
Close and reopen this PR to re-trigger the ci test.

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Thanks for your hard working, I tested all features locally by the command "./vcpkg install skia[*] "but failed:
````
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:158 (message):
    Command failed: "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -C F:/Feature-test/skia/vcpkg/buildtrees/skia/x86-windows-dbg :skia
    Working Directory: F:/Feature-test/skia/vcpkg/buildtrees/skia/src/fa8ee834ed-0fb838c932.clean
    See logs for more information:
      F:\Feature-test\skia\vcpkg\buildtrees\skia\build-x86-windows-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_ninja.cmake:20 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_build_ninja.cmake:41 (z_vcpkg_build_ninja_build)
  installed/x64-windows/share/vcpkg-gn/vcpkg_gn_install.cmake:123 (vcpkg_build_ninja)
  ports/skia/portfile.cmake:279 (vcpkg_gn_install)
  scripts/ports.cmake:145 (include)


Error: Building package skia:x86-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+skia
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?template=report-package-build-failure.md&title=[skia]+Build+error
including:
  package: skia[core,dawn,direct3d,gl,metal,vulkan]:x86-windows -> canvaskit/0.29.0
````
[build-x86-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/8243678/build-x86-windows-dbg-out.log)

**Comment by dave-juicelabs** [comment]
@JonLiu1993 Seems to be the Python 3 build issue again.

**Comment by JonLiu1993** [comment]
@dave-juicelabs, Could you please merge the master? I have no permission to push your branch

**Comment by JonLiu1993** [comment]
@dave-juicelabs ,Is there any progress on this pr?

**Comment by JonLiu1993** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Pull request closed without merging**